.github/workflows/tests.yml

```diff
@@ -0,0 +1,42 @@
+name: Tests
+
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  tests:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.6, 3.7, 3.8]
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - uses: actions/cache@v1
+      with:
+        path: ~/.cache/pip
+        key: ${{ runner.os }}-${{ matrix.python-version }}-pip-${{ hashFiles('**/setup.py') }}
```

Comment:
Does this actually save any significant time? I don't know about GH actions specifically, but I recall one of the CI platforms said that caching typically wasn't worth it for data that's downloaded over a fast link.

Reply:
Doesn't make much of a difference in this case since I don't think there are too many dependencies (just checked and `site-packages` is ~ 320M in my venv for it), but I listened to a podcast a few weeks ago with some pypi people and one of them mentioned their hosting/data transfer costs are like 800,000$+ a month, since then I try to cache pip stuff where possible, especially when running a matrix of tests :joy:

Reply:
Fair enough. Though once it has cached a set of packages, it won't cache any newer versions because the cache key already exists (until we change `setup.py`). I don't know if/when caches are automatically dropped and have to be renewed.

(For the record, Fastly provides free CDN service for PyPI, so the $800k is probably an estimate of what it would cost if they stopped that. It's unlikely the PSF could actually afford that.)

Reply:
Oooooh, okay I feel less bad now :D

That's a good point about the cache not updating when newer packages are released, I guess this is a good time to bring up the option of adding in more restrictive package requirements, which would also let us use dependabot to automatically run tests when our dependencies have new releases.

I'm a fan of the slightly convoluted approach of using Poetry to specify dependencies, then dephell to convert that to a 'traditional' setup.py file ([see this as an example](https://github.com/RobertRosca/oscovida/blob/e430d5cf4b0699b6487e59968cb1be752060966f/setup.py#L45)) which lets you really control what dependencies you have.

But in this case we don't have that many dependencies so it'd be pretty easy to specify them by hand initially and then have dependabot take care of the updates for us. Think we discussed this before a bit, the main (and imo very large) benefit of this is that you'll get an automatic PR with the changes which would then trigger the tests as well, so we can find out if a dependency update caused issues much more easily and consistently.

Reply:
So my take is that the requirements in `setup.py` for a library should be 'abstract requirements', i.e. broad ranges of compatibility, so it can be installed alongside other packages which also require e.g. NumPy.

But maybe it makes sense to run the automated tests with pinned versions of dependencies, for the reasons you describe. It's always annoying when your tests suddenly start failing until you fix a compatibility issue with a new version of something.

Do you know if there's a good way to have abstract dependencies in setup.py and keep a  `test-requirements.txt` file up to date based on that?

Reply:
Yes, it would be possible to have a separate `requirements.txt` file which is used for (1) installing dependencies for the test and (2) used by dependabot to manage provide version bump PRs.

Do you think it would be worth adding that in?